plutus-core/untyped-plutus-core/test/Evaluation/Builtins/Bitwise.hs

```diff
@@ -0,0 +1,564 @@
+-- editorconfig-checker-disable-file
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeApplications  #-}
+
+-- | Tests for [this
+-- CIP](https://github.com/mlabs-haskell/CIPs/blob/koz/bitwise/CIP-XXXX/CIP-XXXX.md)
+module Evaluation.Builtins.Bitwise (
+  shiftHomomorphism,
+  rotateHomomorphism,
+  csbHomomorphism,
+  shiftClear,
+  rotateRollover,
+  csbRotate,
+  shiftPosClearLow,
+  shiftNegClearHigh,
+  rotateMoveBits,
+  csbComplement,
+  csbInclusionExclusion,
+  csbAnd,
+  csbOr,
+  csbXor,
+  ffsReplicate,
+  ffsAnd,
+  ffsOr,
+  ffsXor,
+  ffsIndex
+  ) where
+
+import Control.Monad (unless)
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Evaluation.Builtins.Common (typecheckEvaluateCek, typecheckReadKnownCek)
+import Hedgehog (Property, PropertyT, annotateShow, failure, forAll, forAllWith, property, (===))
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Range qualified as Range
+import Numeric (showHex)
+import PlutusCore qualified as PLC
+import PlutusCore.Evaluation.Machine.ExBudgetingDefaults (defaultBuiltinCostModelForTesting)
+import PlutusCore.MkPlc (builtin, mkConstant, mkIterAppNoAnn)
+import PlutusPrelude (Word8, def)
+import Test.Tasty (TestTree)
+import Test.Tasty.Hedgehog (testPropertyNamed)
+import Test.Tasty.HUnit (assertEqual, assertFailure, testCase)
+import UntypedPlutusCore qualified as UPLC
+
+ffsIndex :: Property
+ffsIndex = property $ do
+  bs <- forAllNonZeroByteString
+  let ffsExp = mkIterAppNoAnn (builtin () PLC.FindFirstSetBit) [
+        mkConstant @ByteString () bs
+        ]
+  case typecheckReadKnownCek def defaultBuiltinCostModelForTesting ffsExp of
+    Left err -> annotateShow err >> failure
+    Right (Left err) -> annotateShow err >> failure
+    Right (Right ix) -> do
+      let hitIxExp = mkIterAppNoAnn (builtin () PLC.ReadBit) [
+            mkConstant @ByteString () bs,
+            mkConstant @Integer () ix
+            ]
+      evaluateAndVerify (mkConstant @Bool () True) hitIxExp
+      unless (ix == 0) $ do
+        i <- forAll . Gen.integral . Range.linear 0 $ ix - 1
+        let missIxExp = mkIterAppNoAnn (builtin () PLC.ReadBit) [
+              mkConstant @ByteString () bs,
+              mkConstant @Integer () i
+              ]
+        evaluateAndVerify (mkConstant @Bool () False) missIxExp
+
+ffsAnd :: Property
+ffsAnd = property $ do
+  bs <- forAllByteString
+  semantics <- forAll Gen.bool
+  let lhs = mkIterAppNoAnn (builtin () PLC.FindFirstSetBit) [
+        mkConstant @ByteString () bs
+        ]
+  let rhsInner = mkIterAppNoAnn (builtin () PLC.AndByteString) [
+        mkConstant @Bool () semantics,
+        mkConstant @ByteString () bs,
+        mkConstant @ByteString () bs
+        ]
+  let rhs = mkIterAppNoAnn (builtin () PLC.FindFirstSetBit) [
+        rhsInner
+        ]
+  evaluateTheSame lhs rhs
+
+ffsOr :: Property
```

Comment:
Unfortunately this doesn't work in general, for two reasons. Firstly, if one of the bytestrings is all-zeroes, this would yield improper answers (`-1` is less than any valid index); secondly, due to how we truncate or pad, this won't work if bytestrings are not of identical lengths.